kudu/src/main/java/org/apache/apex/malhar/kudu/partitioner/KuduOneToManyPartitioner.java

```
@@ -0,0 +1,85 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.malhar.kudu.partitioner;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.apex.malhar.kudu.AbstractKuduInputOperator;
+import org.apache.apex.malhar.kudu.scanner.AbstractKuduPartitionScanner;
+import org.apache.apex.malhar.kudu.scanner.KuduPartitionScanAssignmentMeta;
+import org.apache.apex.malhar.kudu.sqltranslator.SQLToKuduPredicatesTranslator;
+
+/**
+ * Used when a user would like to assign multiple kudu tablets to a single physical instance of the Kudu input operator
+ */
+public class KuduOneToManyPartitioner extends AbstractKuduInputPartitioner
+{
+  private static final Logger LOG = LoggerFactory.getLogger(KuduOneToManyPartitioner.class);
+
+  public KuduOneToManyPartitioner(AbstractKuduInputOperator prototypeOperator)
+  {
+    super(prototypeOperator);
+  }
+
+  /**
+   * Distributes the tablets evenly among the physical instances of the kudu input operator. Please see
+   * {@link AbstractKuduPartitionScanner#preparePlanForScanners(SQLToKuduPredicatesTranslator)} for details as to how
+   * a query based planning is done.
+   * @param totalList The ltotal list of possible tablet scans for all queries
+   * @param context The context that is provided by the framework when repartitioning is to be executed
+   * @return  Map with key as operator id and value as the a list of assignments that can be assigned to that operator.
+   */
+  @Override
+  public Map<Integer, List<KuduPartitionScanAssignmentMeta>> assign(List<KuduPartitionScanAssignmentMeta> totalList,
+      PartitioningContext context)
+  {
+    Map<Integer,List<KuduPartitionScanAssignmentMeta>> partitionAssignments = new HashMap<>();
+    int partitionCount = getNumberOfPartitions(context);
+    if ( partitionCount <= 0) {
```

What causes the partition count to be < 1? If it is a user parameter, it should be a validation error at startup.

---

I have used the approach of letting the user specify the number of apex partitions in two ways;

1. As part of the application config xml ( the usual approach )
2. As part of the code ( setter method )

For the setter method approach, I am having an initial of -1 as the default value to denote that the user did not set anything ( as it can be optional ) . Once the user sets it explicitly , the value is to be used when the partitioning logic is being executed ( thus overriding the app config xml value ).  We could throw an exception in the setter method but having a forced exception at setter level looked a bit ugly. Hence the check when doing the actual planning and forcing it to a minimum value.